### PR TITLE
Implement Data URI imports

### PR DESCRIPTION
**Pull request by goganchic**
Implement feature from #9445

**What kind of change does this PR introduce?**

feature

**Did you add tests for your changes?**

Yes

**Does this PR introduce a breaking change?**

No

**What needs to be documented once your changes are merged?**

Ability to import from data URI.

### CODE REVIEW

**Comment by hems** [comment]
Really interested in this!

Here are some nice articles that are related to this feature and illustrate good use cases:

https://2ality.com/2019/10/eval-via-import.html
https://blog.risingstack.com/writing-a-javascript-framework-sandboxed-code-evaluation/

**Comment by goganchic** [comment]
@sokra would you please take a look at this PR? What should I do to merge it?

File: lib/util/DataURI.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+const decodeDataURI = uri => {
+	const match = URIRegEx.exec(uri);
+	if (!match) return null;
+
+	const isBase64 = match[3];
+	const body = match[4];
+	return isBase64
+		? Buffer.from(body, "base64")
+		: Buffer.from(decodeURIComponent(body), "ascii");
```
Comment (sokra) [inline lib/util/DataURI.js:23]: add a test case with a non-base64 url

File: test/cases/resolving/data-uri/index.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+	const mod = require("data:text/javascript;charset=utf-8;base64,bW9kdWxlLmV4cG9ydHMgPSB7CiAgbnVtYmVyOiA0MiwKICBmbjogKCkgPT4gIkhlbGxvIHdvcmxkIgp9Ow==");
+	expect(mod.number).toBe(42);
+	expect(mod.fn()).toBe("Hello world");
+});
+
+it("should import module from data-uri", function() {
+	const mod = require('./module-with-imports');
+	expect(mod.number).toBe(42);
+	expect(mod.fn()).toBe("Hello world");
+});
```
Comment (sokra) [inline test/cases/resolving/data-uri/index.js:11]: I also would like to see a test case when combining data url with inline loader:

`require("./some-loader!data:text/...")`

Comment (sokra) [reply test/cases/resolving/data-uri/index.js:11]: And a test for json.

**Comment by vankop** [comment]
@goganchic I would like to continue your work if you don't have time for it, so feel free to point me.

**Comment by goganchic** [comment]
> I would be nice to allow matching config rules to data url by mimetype. This is needed to set the correct module types e. g. for json, wasm, assets, etc.
> 
> ```js
> module.rules: [
> {
>   mimetype: "text/javascript",
>   type: "javascript/auto",
>   use: [ ... ]
> },
> {
>   mimetype: "application/wasm",
>   type: "webassembly/async"
> }, 
> {
>   mimetype: "application/json",
>   type: "json"
> }
> ```
> 
> The defaultRules should also include common mimetypes by default.

@sokra do you want to set mimetype for all resources or for Data URI resources only? In the last commit I set mimetype for Data URI resources only.

**Comment by goganchic** [comment]
@sokra would you please take a look at this PR?

File: test/cases/resolving/data-uri/index.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
+it("should require js module from base64 data-uri", function() {
+	const mod = require("data:text/javascript;charset=utf-8;base64,bW9kdWxlLmV4cG9ydHMgPSB7CiAgbnVtYmVyOiA0MiwKICBmbjogKCkgPT4gIkhlbGxvIHdvcmxkIgp9Ow==");
+	expect(mod.number).toBe(42);
+	expect(mod.fn()).toBe("Hello world");
+});
+
+it("should require js module from ascii data-uri", function() {
+	const mod = require("data:text/javascript;charset=utf-8;ascii,module.exports={number:42,fn:()=>\"Hello world\"}");
```
Comment (sokra) [inline test/cases/resolving/data-uri/index.js:8]:
```suggestion
	const mod = require("data:text/javascript;charset=utf-8,module.exports={number:42,fn:()=>\"Hello world\"}");
```

File: lib/NormalModuleFactory.js [hunk 4]
```diff
@@ -474,18 +478,23 @@ class NormalModuleFactory extends ModuleFactory {
...
-					unresolvedResource,
-					resolveContext,
-					(err, resolvedResource, resolvedResourceResolveData) => {
-						if (err) return continueCallback(err);
-						resource = resolvedResource;
-						resourceResolveData = resolvedResourceResolveData;
-						continueCallback();
-					}
-				);
+				if (isDataURI(unresolvedResource)) {
```
Comment (sokra) [inline lib/NormalModuleFactory.js:481]: This matches `unresolvedResource` twice against the data url regexp. Here and for `getMimetype`.

It should be possible to only use `getMimetype` and check `if(mimetype)` here.

Maybe rename to make it clearer: `mimetype` -> `dataUriMimetype`

Comment (goganchic) [reply lib/NormalModuleFactory.js]: fixed

**Comment by sokra** [comment]
Thanks

File: lib/util/DataURI.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
+/*
+	MIT License http://www.opensource.org/licenses/mit-license.php
+*/
+
+"use strict";
+
+// data URL scheme: "data:text/javascript;charset=utf-8;base64,some-string"
+// http://www.ietf.org/rfc/rfc2397.txt
+const URIRegEx = /^data:([^;,]+)?((?:;(?:[^;,]+))*?)(;base64)?,(.*)$/;
```
Comment (evilebottnawi) [inline lib/util/DataURI.js:9]: /cc @goganchic I think we forget `i` flag here, `data` protocol is case insensitivity

Comment (vankop) [reply lib/util/DataURI.js:9]: added in https://github.com/webpack/webpack/pull/11095

**Comment by hems** [comment]
that's some awesome stuff, is that already on a release or RC ?

**Comment by vankop** [comment]
@hems It will shipped with next beta

**Pull request merged**